pkg/controller/azure/service/actuator.go

```diff
@@ -77,43 +73,56 @@ func (a *actuator) CreateOrUpdate(ctx context.Context, obj runtime.Object) (requ
 
 	// Get LoadBalancer IPs
 	ips := getServiceLoadBalancerIPs(svc)
+	shouldIgnore := shouldIgnoreService(svc)
 
 	// Create or update PublicIPAddress objects for existing LoadBalancer IPs
-	for ip := range ips {
-		pubip := &azurev1alpha1.PublicIPAddress{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      generatePublicIPAddressName(svc.Namespace, svc.Name, ip),
-				Namespace: a.namespace,
-			},
-		}
-		a.logger.Info("Creating or updating publicipaddress", "name", pubip.Name, "namespace", pubip.Namespace)
-		if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
-			_, err := controllerutil.CreateOrUpdate(ctx, a.client, pubip, func() error {
-				pubip.Labels = pubipLabels
-				pubip.Spec.IPAddress = ip
-				return nil
-			})
-			return err
-		}); err != nil {
-			return 0, false, errors.Wrap(err, "could not create or update publicipaddress")
+	if !shouldIgnore {
+		for ip := range ips {
+			pubip := &azurev1alpha1.PublicIPAddress{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      generatePublicIPAddressName(svc.Namespace, svc.Name, ip),
+					Namespace: a.namespace,
+				},
+			}
+			a.logger.Info("Creating or updating publicipaddress", "name", pubip.Name, "namespace", pubip.Namespace)
+			if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+				_, err := controllerutil.CreateOrUpdate(ctx, a.client, pubip, func() error {
+					pubip.Labels = pubipLabels
+					delete(pubip.Annotations, azurepublicipaddress.DoNotCleanAnnotation)
+					pubip.Spec.IPAddress = ip
+					return nil
+				})
+				return err
+			}); err != nil {
+				return 0, errors.Wrap(err, "could not create or update publicipaddress")
+			}
 		}
 	}
 
 	// Delete PublicIPAddress objects for non-existing LoadBalancer IPs
 	pubipList := &azurev1alpha1.PublicIPAddressList{}
 	if err := a.client.List(ctx, pubipList, client.InNamespace(a.namespace), client.MatchingLabels(pubipLabels)); err != nil {
-		return 0, false, errors.Wrap(err, "could not list publicipaddresses")
+		return 0, errors.Wrap(err, "could not list publicipaddresses")
 	}
 	for _, pubip := range pubipList.Items {
-		if _, ok := ips[pubip.Spec.IPAddress]; !ok {
+		if _, ok := ips[pubip.Spec.IPAddress]; !ok || shouldIgnore {
```

Comment:
I don't understand the logic here. If the service has the annotation (shouldIgnore) then the pubip will get an annotation but then be immediately deleted. That seems redundant.

Reply:
It's not redundant, in fact it's essential. Unless the annotation is there, the public IP address in Azure will get cleaned up, and this is not what you want. You want the pubip to get deleted without any impact on Azure, and therefore it has to be annotated with the "do-not-clean" annotation before that.